src/App.tsx

```diff
@@ -1,24 +1,246 @@
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<ErrorTypes | null>(null);
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [title, setTitle] = useState<string>('');
+  const [isAdding, setIsAdding] = useState<boolean>(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [selectedTodoIds, setSelectedTodoIds] = useState<number[]>([]);
+  const [isEditing, setIsEditing] = useState<boolean>(false);
+
+  const clearError = () => setError(null);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setAllTodos)
+      .catch(() => {
+        setError(ErrorTypes.Load);
+        setTimeout(clearError, 3000);
+      });
+  }, []);
+
+  const filterHandler = (array: Todo[], filterType: string) => {
+    switch (filterType) {
+      case Filter.Active:
+        return array.filter(item => !item.completed);
+      case Filter.Completed:
+        return array.filter(item => item.completed);
+      default:
+        return array;
+    }
+  };
+
+  const addTodoHandler = useCallback(() => {
+    if (!title.trim()) {
+      setError(ErrorTypes.Empty);
+      setTimeout(clearError, 3000);
+
+      return;
+    }
+
+    setIsAdding(true);
+
+    const todo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      title,
+      userId: USER_ID,
+      completed: false,
+    });
+
+    addTodo(todo)
+      .then(result => {
+        setAllTodos(
+          prevTodos => [...prevTodos, result],
+        );
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setError(ErrorTypes.Add);
+        setTempTodo(null);
+      })
+      .finally(() => {
+        setIsAdding(false);
+        setTitle('');
+        setTempTodo(null);
+      });
+  }, [title]);
+
+  const deleteTodoHandler = useCallback((id: number) => {
+    deleteTodo(id)
+      .then(() => setAllTodos(
+        prevAllTodos => prevAllTodos.filter(prevTodo => prevTodo.id !== id),
+      ))
+      .catch(() => {
+        setError(ErrorTypes.Delete);
+        setTimeout(clearError, 3000);
+      });
+  }, []);
+
+  const activeTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Active), [allTodos],
+  );
+  const completedTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Completed), [allTodos],
+  );
+  const visibleTodos = useMemo(
+    () => filterHandler(allTodos, filter), [allTodos, filter],
+  );
+
+  const deleteAllCompletedHandler = useCallback(() => {
+    completedTodos.forEach(
+      completedTodo => deleteTodoHandler(completedTodo.id),
+    );
+  }, [allTodos]);
+
+  const todoStatusChangeHandler = useCallback((id: number, data: boolean) => {
+    setSelectedTodoIds(prev => [...prev, id]);
+
+    completeTodo(id, data)
+      .then(() => setAllTodos(prevTodos => {
+        const changingTodo = prevTodos.find(todo => todo.id === id);
+
+        if (changingTodo) {
+          changingTodo.completed = data;
```

Comment:
do not mutate initial data ➖

Reply:
I thought that using of "return [...prevTodos]" will prevent from mutation

Reply:
@Volodymyr-Melnyk-Fron10der I thought that using of "return [...prevTodos]" will prevent from mutation

Reply:
using of [...prevTodos] makes copy of array itself but not its elements